xmlrunner/result.py

```diff
@@ -373,21 +380,23 @@ def addSkip(self, test, reason):
         self._save_output_data()
         testinfo = self.infoclass(
             self, test, self.infoclass.SKIP, reason)
+        testinfo.test_exception_name = 'skip'
+        testinfo.test_exception_message = reason
         self.skipped.append((testinfo, reason))
-        self._prepare_callback(testinfo, [], 'SKIP', 'S')
+        self._prepare_callback(testinfo, [], 'skip', 's')
 
     def addExpectedFailure(self, test, err):
         """
         Missing in xmlrunner, copy-pasted from xmlrunner addError.
         """
         self._save_output_data()
 
-        testinfo = self.infoclass(self, test, self.infoclass.ERROR, err)
-        testinfo.test_exception_name = 'ExpectedFailure'
-        testinfo.test_exception_message = 'EXPECTED FAILURE: {}'.format(testinfo.test_exception_message)
+        testinfo = self.infoclass(self, test, self.infoclass.SKIP, err)
```

Comment:
https://docs.python.org/3/library/unittest.html#unittest.expectedFailure

> Mark the test as an expected failure. If the test fails it will be considered a success. If the test passes, it will be considered a failure.

Can you please elaborate **why** it should be marked as `SKIP` instead of `ERROR` or `FAILURE`?
To me a failure is a failure, expected or not; so I'm not sure I understand the change.

Reply:
Can you actually split the changes and address this semantic change separately? I'm okay merging the rest.

Reply:
I'll say that for Django's test suite, a recent change in unittest-xml-reporting
 is making Jenkins marking our builds as "Unstable" rather than "Success" due to tests marked with `@unittest.expectedFailure` failing (as exected).

Reply:
@timgraham, could be unrelated to this PR but related to some of my changes. please file a ticket with more info... I don't see `django/django` in travis, so I'm not sure which CI is used.

Reply:
What I meant is that I think this change would fix that issue. Possibly the regression is due to cc0567933171aed91b9587941f8d43293f394be7. Our CI is at https://djangoci.com/.

Reply:
> https://docs.python.org/3/library/unittest.html#unittest.expectedFailure
> 
> > Mark the test as an expected failure. If the test fails it will be considered a success. If the test passes, it will be considered a failure.
> 
> Can you please elaborate **why** it should be marked as `SKIP` instead of `ERROR` or `FAILURE`?
> To me a failure is a failure, expected or not; so I'm not sure I understand the change.

An expected failure is a way of marking as “incorrect” — a good example might be a test that is in fact correct, but which tests functionality not currently working. The entire point of an expected failure is to ensure that the test suite remains passing while semantically marking a test as “wrong” or “broken”.

Compared to the obvious alternative of skipping the test, the main feature of `expectedFailure` is that the test run _actually fails_ if the test suddenly starts working, preventing it from happening accidentally. From the [CPython 3.7](https://github.com/python/cpython/blob/v3.7.0/Lib/unittest/result.py#L160) sources:

```python
    def wasSuccessful(self):
        """Tells whether or not this result was a success."""
[snip]
        return ((len(self.failures) == len(self.errors) == 0) and
                (not hasattr(self, 'unexpectedSuccesses') or
                 len(self.unexpectedSuccesses) == 0))
```

Given that the JUnit XML does not support this feature, a “skip” is reasonably close approximation, although one might also mark it as a success. The latter feels a bit wrong to me. An unexpected success, however, is a clear failure.

As with Django, the motivation for me was that our test suite suddenly started becoming “unstable” on Jenkins — I'm suspect that wasn't the case with the code I originally submitted, but you never know…